File: simplebase/src/main/java/com/zjh/simplebase/base/BaseViewModel.java
package com.zjh.simplebase.base;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;

/**
 * base viewModel
 *
 * @author zhujianhua
 * on 2021/1/6
 */
public class BaseViewModel extends AndroidViewModel implements IBaseViewModel {

    public BaseViewModel(@NonNull Application application) {
        super(application);
    }
}

File: simplebase/src/main/java/com/zjh/simplebase/base/BaseViewModel.kt
package com.zjh.simplebase.base

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * base viewModel
 * @author zjh
 * on 2021/5/13
 */
open class BaseViewModel(application: Application) : AndroidViewModel(application), IBaseViewModel {

    /**
     * 加载中
     */
//    val loadingLiveData = SingleLiveData<Boolean>()

    /**
     * 接口请求
     */
    fun requestLaunch(
            //加载、处理数据
            block: suspend () -> Unit,
            //异常处理
            error: suspend (Throwable) -> Unit,
            //结束操作 非必传
            completed: (suspend () -> Unit)? = null
    ) {
//        loadingLiveData.postValue(true)
        viewModelScope.launch(Dispatchers.IO) {
            try {
                block()
            } catch (e: Exception) {
                //404 或者其他特殊情况
                error(e)
            } finally {
                //结束
                completed?.let { it() }
            }
        }
    }

}